spdx/v3/v3_0/spdx_test.go

```diff
@@ -0,0 +1,105 @@
+package v3_0_test
+
+import (
+	"testing"
+
+	spdx "github.com/spdx/tools-golang/spdx/v3/v3_0"
+)
+
+func Test_exportImportExport(t *testing.T) {
+	// create a document
+	doc := spdx.NewDocument(
+		spdx.ProfileIdentifierType_Software,
+		"My Document",
+		&spdx.SoftwareAgent{Agent: spdx.Agent{Element: spdx.Element{
+			Name:    "tools-golang",
+			Summary: "a summary",
+		}}},
+		&spdx.Tool{Element: spdx.Element{
+			ExternalIdentifiers: spdx.ExternalIdentifierList{
+				&spdx.ExternalIdentifier{
+					ExternalIdentifierType: spdx.ExternalIdentifierType_Cpe23,
+					Identifier:             "cpe:2.3:a:myvendor:my-product:*:*:*:*:*:*:*:*",
+				},
+			},
+			ExternalRefs: nil,
+			Name:         "not-tools-golang",
+		}})
+
+	sbom := &spdx.SoftwareSbom{}
+	doc.RootElements = append(doc.RootElements, sbom)
+
+	// create a package
+
+	pkg1 := &spdx.SoftwarePackage{
+		SoftwareSoftwareArtifact: spdx.SoftwareSoftwareArtifact{Artifact: spdx.Artifact{Element: spdx.Element{
+			Name: "some-package-1",
+		}}},
+		SoftwarePackageVersion: "1.2.3",
+	}
+
+	// create another package
+
+	pkg2 := &spdx.AiAIPackage{}
+	pkg2.Name = "some-package-2"
+	pkg2.SoftwarePackageVersion = "2.4.5"
+
+	// add the packages to the sbom
+
+	sbom.RootElements = append(sbom.RootElements, pkg1, pkg2)
+
+	// add a file
+
+	file1 := &spdx.SoftwareFile{SoftwareSoftwareArtifact: spdx.SoftwareSoftwareArtifact{Artifact: spdx.Artifact{Element: spdx.Element{
+		Name: "/bin/bash",
+	}}}}
+	sbom.RootElements = append(sbom.RootElements, file1)
+
+	// add relationships
+
+	sbom.RootElements = append(sbom.RootElements, &spdx.Relationship{
+		From:             file1,
+		RelationshipType: spdx.RelationshipType_Contains,
+		Tos: spdx.ElementList{
+			pkg1,
+			pkg2,
+		},
+	})
+
+	sbom.RootElements = append(sbom.RootElements, &spdx.Relationship{
+		From:             pkg1,
+		RelationshipType: spdx.RelationshipType_DependsOn,
+		Tos: spdx.ElementList{
+			pkg2,
+		},
+	})
+
+	// some basic usage:
+
+	var pkgs []*spdx.SoftwarePackage
+	for _, sbom := range doc.RootElements.SoftwareSbomIter() {
+		for _, rel := range sbom.RootElements.RelationshipIter() {
+			if rel.RelationshipType != spdx.RelationshipType_Contains {
+				continue
+			}
+			_ = spdx.As(rel.From, func(f *spdx.SoftwareFile) any {
```

Comment:
The second thing is accessing single-instance properties like this. There is no magic typing as is done with the specialized collections that could be done to add methods like rel.From.SoftwareFile(), for example. This means it's required to type assert... but because of the inheritance, you cannot type assert to *spdx.SoftwareFile, or you'll miss any subtypes, so the spdx.As function is a helper that simply switches on the type provided -- in this case *spdx.SoftwareFile switches to test for spdx.AnySoftwareFile interface, and if found calls .asSoftwareFile() to get a reference. This allows the inheritance to work properly at the expense of not really being able to do anything idiomatic here.